modules/core/src/smithy4s/schema/Alt.scala

```diff
@@ -36,7 +37,14 @@ final case class Alt[F[_], U, A](
 }
 object Alt {
 
-  final case class WithValue[F[_], U, A](alt: Alt[F, U, A], value: A) {
+  @nowarn
+  final case class WithValue[F[_], U, A](
+      @deprecated(
```

Comment:
In this PR, I'm deprecating access to the `alt` val, as it requires implementors to create their own caching mechanism, which is inherently risky due to how schemas are not stable (you need to cache using labels as keys). 

A fair chunk of this PR is a consequence of this deprecation, as interpreters/compilers are amended to make use of `Alt.Dispatcher` instead, which takes care of memoization for the implementor in the correct way, and offers a more elegant way to implement encoders of unions, by hiding the ugliness.